### PR TITLE
Fix Up FileManager Tests on Windows

### PR DESCRIPTION
**Pull request by gmittert**
- Windows directory enumerator now doesn't look in a directory it
  returns until nextObject is called again
- Fixed some path handling platform issues
- Properly calculate level

See SR-10738

### CODE REVIEW

**Comment by gmittert** [comment]
@spevans Is there a particular scenario you're testing for with the file names ending with `.` (https://github.com/apple/swift-corelibs-foundation/pull/2277/files#diff-fffbc0319b30adac9f635327a750b82fR509)? Windows isn't capable of create/working with files ending with `.` so I'm trying to figure out what I want to do with those test cases on Windows.

File: Foundation/NSPathUtilities.swift [hunk 1]
```diff
@@ -345,7 +349,11 @@ extension NSString {
         
         let automount = "/var/automount"
         resolved = resolved._tryToRemovePathPrefix(automount) ?? resolved
+#if os(Windows)
+        return String(resolved.map { $0 == "\\" ? "/" : $0 })
```
Comment (compnerd) [inline Foundation/NSPathUtilities.swift]: Why change the path separator here?

File: Foundation/NSURL.swift [hunk 2]
```diff
@@ -32,7 +32,11 @@ private func _standardizedPath(_ path: String) -> String {
     if !path.absolutePath {
         return path._nsObject.standardizingPath
     }
+    #if os(Windows)
+    return String(path.map({$0 == "/" ? "\\" : $0}))
```
Comment (compnerd) [inline Foundation/NSURL.swift]: My understanding is that the standardizing should collapse relative components (i.e. `..` should be chased, `.` dropped).  As such, any reason to not just use `PathCchCanonicalizeEx`?

File: Foundation/NSPathUtilities.swift [hunk 3]
```diff
@@ -118,7 +118,11 @@ extension String {
     }
 
     internal var absolutePath: Bool {
```
Comment (compnerd) [inline Foundation/NSPathUtilities.swift]: Bleh!  It would be nice to rename this to `isAbsolutePath`.

File: TestFoundation/TestFileManager.swift [hunk 4]
```diff
@@ -447,13 +447,25 @@ class TestFileManager : XCTestCase {
...
+                } else if item == "path2\\item" {
+                    item2FileAttributes = e.fileAttributes
+                }
+#else
                 if item == "item" {
                     item1FileAttributes = e.fileAttributes
                 } else if item == "path2/item" {
                     item2FileAttributes = e.fileAttributes
                 }
+#endif
```
Comment (compnerd) [inline TestFoundation/TestFileManager.swift]: Isn't this hunk equivalent to just changing L459 (R) to `} else if item == "path2/item" || item == "path2\\item" {`?

File: TestFoundation/TestFileManager.swift [hunk 5]
```diff
@@ -447,13 +447,25 @@ class TestFileManager : XCTestCase {
...
             }
+#if os(Windows)
+            XCTAssertEqual(foundItems, Set(["item", "path2", "path2\\item"]))
```
Comment (compnerd) [inline TestFoundation/TestFileManager.swift]: I wonder if you can span the `Set` over two lines and do an `#if os(Windows)` to swap out the third element.

File: Foundation/FileManager+Win32.swift [hunk 6]
```diff
@@ -689,27 +690,52 @@ extension FileManager {
...
-                               _NSErrorWithWindowsError(GetLastError(), reading: true))
-                           if !keepGoing { return nil }
-                        }
-                        continue
-                    }
-                    _stack.append(contentsOf: dirContents)
+
+            func findFirstValidFile() -> URL? {
+              while let url = _stack.popLast() {
+                let dirFSR = url.withUnsafeFileSystemRepresentation { $0.flatMap { fsr in String(utf8String: fsr) } } ?? ""
```
Comment (compnerd) [inline Foundation/FileManager+Win32.swift]: Hmm, this seems pretty expensive.

File: Foundation/FileManager+Win32.swift [hunk 7]
```diff
@@ -689,27 +690,52 @@ extension FileManager {
...
+                if !FileManager.default.fileExists(atPath: dirFSR, isDirectory: nil) {
```
Comment (compnerd) [inline Foundation/FileManager+Win32.swift]: Huh?  Should `dirFSR` be renamed?  If `isDirectory` is `nil`, then we do not know if it is a file or directory, so it seems to be misleading.

File: Foundation/FileManager+Win32.swift [hunk 8]
```diff
@@ -689,27 +690,52 @@ extension FileManager {
...
                 _current = url
-                return url
+                return _current
+              }
+              return nil
             }
-            return nil
+
+            var isDir: ObjCBool = false
+            let dirFSR = _current.withUnsafeFileSystemRepresentation { $0.flatMap { fsr in String(utf8String: fsr) } } ?? ""
```
Comment (compnerd) [inline Foundation/FileManager+Win32.swift]: Again, the `dirFSR` is really misleading here.  Can we rename this as we do not know what the path points to.

**Comment by gmittert** [comment]
- Cleaned up the directory enumerator code to be less terrible
- Fixed some slash issues and removed the extra conversions

**Comment by gmittert** [comment]
@swift-ci please test

**Comment by gmittert** [comment]
@swift-ci please test

File: CoreFoundation/URL.subproj/CFURL.c [hunk 9]
```diff
@@ -4474,7 +4474,7 @@ CFStringRef CFURLCreateStringWithFileSystemPath(CFAllocatorRef allocator, CFURLR
     //	and do a linked-on-or-later check so we don't break third parties.
     //	See <rdar://problem/4003028> Converting volume name from POSIX to HFS form fails and
     //	<rdar://problem/4018895> CF needs to back out 4003028 for icky details.
-    if ( relPath && CFURLHasDirectoryPath(anURL) && CFStringGetLength(relPath) > 1 && CFStringGetCharacterAtIndex(relPath, CFStringGetLength(relPath)-1) == '/') {
+    if ( relPath && CFURLHasDirectoryPath(anURL) && CFStringGetLength(relPath) > 1 && CFStringGetCharacterAtIndex(relPath, CFStringGetLength(relPath)-1) == _CFGetSlash()) {
```
Comment (compnerd) [inline CoreFoundation/URL.subproj/CFURL.c]: Since this is from a URL, is this change needed, and if it is from a URL is this not incorrect?  CC: @millenomi

File: TestFoundation/TestFileManager.swift [hunk 10]
```diff
@@ -492,10 +497,17 @@ class TestFileManager : XCTestCase {
         let fm = FileManager.default
         let basePath = NSTemporaryDirectory() + "testdir\(NSUUID().uuidString)/"
         let subDirs1 = basePath + "subdir1/subdir2/.hiddenDir/subdir3/"
-        let subDirs2 = basePath + "subdir1/subdir2/subdir4.app/subdir5./.subdir6.ext/subdir7.ext./"
         let itemPath1 = basePath + "itemFile1"
+#if os(Windows)
+        // Filenames ending with '.' are not valid on Windows, so don't bother testing them
```
Comment (compnerd) [inline TestFoundation/TestFileManager.swift]: @millenomi - what do you think about this?

File: CoreFoundation/URL.subproj/CFURL.c (hunk 9 above)
Comment (gmittert) [reply CoreFoundation/URL.subproj/CFURL.c]: Seems like it's not, I've removed that change.

File: TestFoundation/TestFileManager.swift (hunk 5 above)
Comment (millenomi) [reply TestFoundation/TestFileManager.swift]: You can't. You can move the set creation elsewhere, or make it mutable and `#if os(Windows)` the addition.

**Comment by gmittert** [comment]
@swift-ci please test

File: TestFoundation/TestFileManager.swift (hunk 10 above)
Comment (millenomi) [reply TestFoundation/TestFileManager.swift]: That's fine.

**Comment by millenomi** [comment]
@swift-ci please test and merge

**Comment by millenomi** [comment]
@swift-ci please test and merge
